File: go/0077-combinations.go
func backtrack(n, k, start int, arr []int, ans *[][]int) {
	if len(arr) == k {
		comb := make([]int, k)
		copy(comb, arr)
		*ans = append(*ans, comb)
		return
	}
	for i := start; i <= n-k+(len(arr)+1); i++ {
		arr = append(arr, i)
		backtrack(n, k, i+1, arr, ans)
		arr = arr[:len(arr)-1]
	}
}

func combine(n int, k int) [][]int {
	ans := [][]int{}
	backtrack(n, k, 1, []int{}, &ans)
	return ans
}

File: cpp/0077-combinations.cpp
class Solution {
private:
    void backtrack(int start, int n, int k, vector<int> &combination, vector<vector<int>> &res){
            //base case, when size of combination is k, we wanna stop
            if(combination.size() == k){
                res.push_back(combination);
                return;
            }

            for(int i = start; i<=n; i++){
                combination.push_back(i);
                backtrack(i+1, n, k, combination, res);
                combination.pop_back();
            }
        }
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;

        //initial empty list to pass to the backtrack function
        vector<int> emptyCombination;

        backtrack(1, n, k, emptyCombination, res);

        return res;
    }
};